### ISSUE TITLE
CLI feature: support stdin

### ISSUE THREAD

**Issue by emmanueltouzery** [checkstyle/checkstyle#14698]
It would be nice if the CLI could allow giving in an input file to check through stdin rather than giving a filename. That would allow linting as the user types even unsaved files without going through temporary files.

This is a super useful project otherwise!

For instance, where a CLI invocation today would be:
`java -jar checkstyle-10.14.2-all.jar -c /sun_checks.xml MyClass.java`

I'd like to see:
`cat MyClass.java | java -jar checkstyle-10.14.2-all.jar -c /sun_checks.xml -`

in this case I assumed the parameter `-` would signal to read the file to check from stdin, but any other parameter could do. For instance, `eslint` uses a `--stdin` flag for that purpose.

The advantage is that it's then possible to lint a file without saving it, typically you can lint "as you type" in an editor without having to save the edited file to disk in a temporary file just to be able to invoke checkstyle.

Since checkstyle supports several file types (java, properties..), maybe a parameter `--stdin-filename` (that's the eslint flag name) would allow to discriminate which file type to take into account.

**Comment by rnveach** [comment]
Can you update your post and give an example command line of what you are looking for.

We use https://picocli.info/ , so what can and only support whatever they provide us.

**Comment by emmanueltouzery** [comment]
I have updated the issue description!

**Comment by romani** [comment]
@remkop , I went through doc I didn't not find this possibility explicitly.
Do you know how to archive something like user requesting above?

**Comment by remkop** [comment]
Hi @romani, yes, this is possible, but the picocli library does not offer special support for this.
This is up to the application to handle.

Basically, when the end user specifies `-` as a file name, the application should not open a `FileInputStream` from the specified path, but instead should start reading from STDIN, until EOF is reached.

Looking at the CheckStyle code, 
the `CliOptions` class has this for positional parameters:

```java
/** List of file to validate. */
@Parameters(arity = "1..*", description = "One or more source files to verify")
private List<File> files;
```

Depending on the other options, the actual processing is delegated to various other classes, like `AstTreeStringPrinter.printFileAst`, `XpathUtil.printXpathBranch`, `DetailNodeTreeStringPrinter.printFileAst`, etc.

So perhaps all these places need to be changed.
But I notice that CheckStyle already has an abstraction `FileText` that seems to be responsible for extracting the contents from a `File` object. 

It appears to me that the `FileText` class is a good candidate to add this logic: read from STDIN instead of from the specified path, if the specified path is equal to the string `"-"`.

**Comment by remkop** [comment]
I added an example to `picocli-examples`: 
https://github.com/remkop/picocli/blob/main/picocli-examples/src/main/java/picocli/examples/stdin/PrintFirstLine.java

**Comment by romani** [comment]
@remkop , thanks a lot for all very detailed comments. Very helpful.

@emmanueltouzery ,

> Since checkstyle supports several file types (java, properties..), maybe a parameter --stdin-filename

after a bit of thoughts on this .... there are Checks that validate filename against top class, and there are something for folders that contains a target file, for property file validation (TranslationCheck) it even need multiple files.
This feature might work and but it will be very prone to false positives in general usage, dependeds with what  config(set of Checks) you will run it.

you can use suggestion above and hack some solution and we can consider it for review. But it looks like it will be much easier to user content to temp file and run existing CLI, as you know context of execution, you might be able to filter out false positives that this solution prone to.

> That would allow linting as the user types even unsaved files without going through temporary files.

One more problem is file has to be compile-able. https://checkstyle.org/writingchecks.html#Limitations
`Code has to be compilable by javac to get valid violations. If it is not, you can get hard to understand parse errors.` sometime is works on bad java files but Checks are go wild. Compilation with no errors is hard limitation for us.
This point does not a blocker for this issue, it might be a problem for task what you try solve. `javac` must pass before you put file to checkstyle CLI.

**Comment by emmanueltouzery** [comment]
Right. Thank you for taking the time to research all the ins and outs of the idea. I understand all the constraints now.
I didn't realize that some checks go and check other files on disk... That seems to be the biggest hurdle, because it would have to be tracked, which checks are compatible, which not, added complexity...

Ok, then I'll close this issue for now, I'll consider whether to make a PR, but given all the constraints you've quoted, I likely won't... Thank you again for all the effort in checking this out, and for the tool itself which is really super useful!
